File: 2023/day7/main.py
from collections import Counter
from functools import partial


LABELS = ('2', '3', '4', '5', '6', '7', '8', '9', 'T', 'J', 'Q', 'K', 'A')
LABELS_WITH_JOKER = ('J', '2', '3', '4', '5', '6', '7', '8', '9', 'T', 'Q', 'K', 'A')


def get_hand_value(hand: str, label_to_value: dict[str, int]) -> int:
    """Treat hand as a number of base len(label_to_value)"""
    value = 0
    base = len(label_to_value)

    for i, card in enumerate(reversed(hand)):
        value += label_to_value[card] * base ** i

    return value


def replace_jokers(hand: str, default_replacement: str = '2'):
    counter = Counter(hand)
    most_common = counter.most_common(2)
    label, count = most_common[0]
    # If 'J' is the most common use the 2nd most common
    if label == 'J':
        # If all cards are 'J's replace with '2'
        if count == 5:
            label = default_replacement
        else:
            label, _ = most_common[1]
    return hand.replace('J', label)


def rank_hand(
        hand_and_bid: list[str, str],
        labels: tuple[str] = LABELS,
        should_replace_jokers: bool = False
) -> int:
    hand, _ = hand_and_bid
    label_to_value = dict(zip(labels, range(len(labels))))
    value = get_hand_value(hand, label_to_value)

    if should_replace_jokers:
        hand = replace_jokers(hand)

    counter = Counter(hand)
    n_labels = len(counter)
    counts = counter.values()

    hand_type = None
    if n_labels == 1:
        hand_type = 6
    elif n_labels == 2 and set(counts) == {4, 1}:
        hand_type = 5
    elif n_labels == 2 and set(counts) == {3, 2}:
        hand_type = 4
    elif n_labels == 3 and set(counts) == {3, 1, 1}:
        hand_type = 3
    elif n_labels == 3 and set(counts) == {2, 2, 1}:
        hand_type = 2
    elif n_labels == 4:
        hand_type = 1
    else:
        hand_type = 0

    # Add the hand type as the most significant digit
    return value + label_to_value[labels[hand_type]] * len(labels) ** len(hand)


def main():
    with open("input.txt") as f:
        hands_and_bids = [line.split() for line in f.read().splitlines()]

    s1 = 0
    for i, (hand, bid) in enumerate(
        sorted(hands_and_bids, key=rank_hand), start=1
    ):
        s1 += i * int(bid)

    assert s1 == 253866470
    print(f"{s1=}")

    s2 = 0
    key_fn = partial(
        rank_hand, labels=LABELS_WITH_JOKER, should_replace_jokers=True
    )
    for i, (hand, bid) in enumerate(
        sorted(hands_and_bids, key=key_fn), start=1
    ):
        s2 += i * int(bid)

    assert s2 == 254494947
    print(f"{s2=}")


if __name__ == '__main__':
    main()

    # # test it doesn't do any replacement if there are no 'J's
    # assert replace_jokers('23456') == '23456'
    # # test it chooses the most common as replacement for 'J's
    # assert replace_jokers('KTJJT') == 'KTTTT'
    # # test one replacement
    # assert replace_jokers('2345J') == '23452'
    # # test replace all
    # assert replace_jokers('JJJJJ') == '22222'
    # assert replace_jokers('JJJJJ', default_replacement='3') == '33333'

File: 2023/day7/main.go
package main

import (
	"bufio"
	"cmp"
	"fmt"
	"log"
	"math"
	"os"
	"slices"
	"strconv"
	"strings"
)

type Hand struct {
	Cards string
	Bid   int
}

func countRunes(s string) map[rune]int {
	counter := make(map[rune]int)
	for _, label := range s {
		if count, _ := counter[label]; true {
			counter[label] = count + 1
		}
	}
	return counter
}

func getCardsValue(h Hand, labelToValue map[rune]int) int {
	/*
		Treat h.Cards as a number in a numeric system with base len(labelToValue)
	*/
	value := 0
	mostSignificantIdx := len(h.Cards) - 1
	base := len(labelToValue)
	for i, label := range h.Cards {
		cardValue, _ := labelToValue[label]
		value += cardValue * int(math.Pow(float64(base), float64(mostSignificantIdx-i)))
	}
	return value
}

func replaceJokers(cards string, defaultReplacement rune) string {
	/*
		Find if there are 'J's among the cards and replace them with the best option
	*/

	if cards == "JJJJJ" {
		return strings.ReplaceAll(cards, "J", string(defaultReplacement))
	}

	// Find the most used label
	counter := countRunes(cards)
	max := -1
	var labelReplacement rune
	for label, count := range counter {
		if label != 'J' && count > max {
			labelReplacement = label
			max = count
		}
	}

	if labelReplacement != 0 {
		return strings.ReplaceAll(cards, "J", string(labelReplacement))
	}

	return cards
}

func getHandValue(h Hand, labels []rune, shouldReplaceJokers bool) int {
	labelToValue := make(map[rune]int)
	for i, l := range labels {
		labelToValue[l] = i
	}

	cardsValue := getCardsValue(h, labelToValue)

	cards := h.Cards
	if shouldReplaceJokers {
		cards = replaceJokers(cards, labels[1])
	}
	cardCounter := countRunes(cards)
	lenCardCounter := len(cardCounter)

	var handType int
	if lenCardCounter == 1 {
		handType = 6
	} else if lenCardCounter == 2 {
		for _, value := range cardCounter {
			if value == 1 || value == 4 {
				handType = 5
				break
			} else {
				handType = 4
				break
			}
		}
	} else if lenCardCounter == 3 {
		for _, value := range cardCounter {
			if value == 3 {
				handType = 3
				break
			} else if value == 2 {
				handType = 2
				break
			}
		}
	} else if lenCardCounter == 4 {
		handType = 1
	}

	handTypeValue := labelToValue[labels[handType]]
	// Add handTypeValue as the most significant digit
	return cardsValue + handTypeValue*int(math.Pow(float64(len(labels)), float64(len(h.Cards))))
}

func main() {
	f, err := os.Open("input.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()

	scanner := bufio.NewScanner(f)
	var hands []Hand

	for scanner.Scan() {
		line := scanner.Text()
		cardsAndBid := strings.Split(line, " ")
		cards, bidStr := cardsAndBid[0], cardsAndBid[1]
		bid, err := strconv.Atoi(bidStr)
		if err != nil {
			log.Fatal(err)
		}
		hands = append(hands, Hand{Cards: cards, Bid: bid})
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	cardLabels := [13]rune{'2', '3', '4', '5', '6', '7', '8', '9', 'T', 'J', 'Q', 'K', 'A'}
	slices.SortFunc(hands, func(a, b Hand) int {
		return cmp.Compare(getHandValue(a, cardLabels[:], false), getHandValue(b, cardLabels[:], false))
	})

	s1 := 0
	for i, h := range hands {
		s1 += h.Bid * (i + 1)
	}
	fmt.Println("s1:", s1)

	cardLabelsWithJoker := [13]rune{'J', '2', '3', '4', '5', '6', '7', '8', '9', 'T', 'Q', 'K', 'A'}
	slices.SortFunc(hands, func(a, b Hand) int {
		return cmp.Compare(getHandValue(a, cardLabelsWithJoker[:], true), getHandValue(b, cardLabelsWithJoker[:], true))
	})

	s2 := 0
	for i, h := range hands {
		s2 += h.Bid * (i + 1)
	}
	fmt.Println("s2:", s2)

	// fmt.Println(replaceJokers("23456", '3')) // "23456"
	// fmt.Println(replaceJokers("KTJJT", '3')) // "KTTTT"
	// fmt.Println(replaceJokers("2345J", '3')) // "23455"
	// fmt.Println(replaceJokers("JJJJJ", '2')) // "22222"
	// fmt.Println(replaceJokers("JJJJJ", '3')) // "33333"
}
